### PR TITLE
Don't copy unneeded grads when using a function for several derivatives

### LINKED ISSUE #7722: Error in backprop when using frozen LSTM layers (new with 0.4.0)
## Issue description

The general situation is that we have a pretrained Language Model, and during a first phase we only want to train the new embedding layer we added before fine-tuning the whole thing. This worked fine in 0.3 but now sends an error message during back propagation. Minimal reproduction is to just create a simple model with a linear layer and an LSTM, freeze this second layer (by applying require_grads=False to its parameters) and try to compute a back-propagation.

## Code example

See [here](https://github.com/sgugger/Deep-Learning/blob/master/Bug%20with%20frozen%20LSTM%20layer.ipynb)

## System Info
PyTorch version: 0.4.0                                                                                                  
Is debug build: No                                                                                                      
CUDA used to build PyTorch: 9.0                                                                                                                                                                                                                 OS: Microsoft Windows 10 Home                                                                                           
GCC version: Could not collect                                                                                         
CMake version: Could not collect                                                                                                                                                                                                                Python version: 3.6                                                                                                     
Is CUDA available: Yes                                                                                                  
CUDA runtime version: 9.1.85                                                                                            
GPU models and configuration: Could not collect                                                                         
Nvidia driver version: Could not collect                                                                                
cuDNN version: Could not collect                                                                                                                                                                                                                Versions of relevant libraries:                                                                                         
[pip] Could not collect                                                                                                 
[conda] Could not collect

Pytorch was installed with conda, the bug also appears on my linux instances.

Thanks for your help!

### PR DESCRIPTION
**Pull request by t-vi**
Fixes #7722 

Trying to copy all results fails when one of them is a tensor list which
has not been populated. This blew up for CuDNN RNNs when the weights
did not require grad.

Thanks to Sylvain Gugger for reporting!

### CODE REVIEW

File: tools/autograd/gen_autograd_functions.py [hunk 1]
```diff
@@ -63,6 +63,12 @@
 }
 """)
 
+DERIVATIVE_MULTI_COPY_RANGE = CodeTemplate("""\
+  if (should_compute_output({ ${name}_ix })) {
+    copy_range(grad_inputs, ${name}_ix, std::get<${i}>(grad_result));
+  }
```
Comment (apaszke) [inline tools/autograd/gen_autograd_functions.py]: Wouldn't this cause some of the returned gradients to be undefined? I think we have a pretty strict no-undefined policy in autograd, so we might still allocate zero inputs. Am I right @colesbury?

**Review by apaszke — DISMISSED** [review]

File: tools/autograd/gen_autograd_functions.py (hunk 1 above)
Comment (t-vi) [reply tools/autograd/gen_autograd_functions.py]: From inspecting the generated Functions.cpp I had the impression that the end result when should_compute_output now is exactly the same as it were when the other grads of the "MULTI" had should_compute_output.
I must admit that I still cannot figure out what to change when revisiting Functions.cpp with your comment in mind.

Comment (t-vi) [reply tools/autograd/gen_autograd_functions.py]: Oh, the above is hard to read without the context of the actual code. The generated snippet for cudnn rnn backward *before* the PR is:
```
  if (should_compute_output({ input_ix, hx_ix, cx_ix, weight_ix })) {
      auto grad_input_mask = std::array<bool, 4>{
      should_compute_output({ input_ix }),
      should_compute_output({ hx_ix }),
      should_compute_output({ cx_ix }),
      should_compute_output({ weight_ix }),
    };
    auto grad_result = _cudnn_rnn_backward(input, weight, weight_stride0, result4, hx, cx, result0, grads[0], grads[1], grads[2], mode, hidden_size, num_layers, $
    copy_range(grad_inputs, input_ix, std::get<0>(grad_result));
    copy_range(grad_inputs, hx_ix, std::get<1>(grad_result));
    copy_range(grad_inputs, cx_ix, std::get<2>(grad_result));
    copy_range(grad_inputs, weight_ix, std::get<3>(grad_result));
  }
```
and afterwards, you get the additional `if`s around the copy_range calls.
Looking at that, I might use `grad_input_mask` instead of `should_compute_output` if you prefer.

Comment (ezyang) [reply tools/autograd/gen_autograd_functions.py]: Yes, this looks right to me. If we are not requested to compute output for something, we're allowed to return undefined.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@apaszke Can you take a second look and see if you agree with the reasoning? I'd prefer not to override your concern.

**Comment by apaszke** [comment]
I’m 99% sure we’re not supposed to ever pass around undefined tensors. On the other hand, if should_compute_output checks that next function is a nullptr, then it will probably be discarded anyway, so it should be ok. I guess it’s safe to merge now that undefined tensors can’t cause memory corruption

**Comment by ezyang** [comment]
My understanding of the invariant is documented at https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#undefined-tensor-conventions
